skimage/io/util.py

```diff
@@ -28,7 +29,15 @@ def file_or_url_context(resource_name):
                 f.write(u.read())
             # f must be closed before yielding
             yield f.name
-        finally:
+        except (URLError, HTTPError):
+            # could not open URL
+            os.remove(f.name)
+            raise
+        except (FileNotFoundError, FileExistsError,
```

Comment:
Shouldn't this be a catch-all?

Reply:
Perhaps having BaseException makes it so?

Reply:
BaseException probably makes it so, but that is only there because it is specifically raised by NamedTemporaryFile. Perhaps you are right that it should be catch-all, as there might be things called by NamedTemporaryFile that might raise in other, unexpected ways. Either way, I think catch-all should be a separate clause.

Reply:
I'd usually agree, except that the existing clause is empty, so it doesn't make much of a difference.  Either way is fine, though.

Reply:
Actually, here's the reason I don't like a bare clause: in the case of an unexpected exception, it's unclear whether we should `os.remove(f.name)`, because it's unclear whether the exception was raised by `NamedTemporaryFile()`, `urlopen`, or `f.write`. So, I'd rather not add an `except` clause...

One possible option:

```
finally:
    if 'f' in locals() and os.path.exists(f):
        os.remove(f.name)
```

But, this still feels brittle. Overall I think we should just leave this be until someone complains. =)

Reply:
Sounds then like another try/except is required to specialize what is being caught.